### PR TITLE
chore: Switch useSettings to observable [WEB-801]

### PR DESCRIPTION
**Pull request by gt2345**
## Description

We want to use observables to persist state for useSettings instead of react context.

## Test Plan

Navigate to a page related to settings, like task list page.
Change settings like table sort key.
Refresh the page, settings should persist.
Open the same page under incognito mode, log in as the same user, settings should persist.  

On task list page, select couple filters.
Filtered value should show up in the url.
Reset filters, filtered value should disappear from the url.

While changing settings, open network tab and verify that no duplicated API calls have made to `/setting` endpoint

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[WEB-801](https://determinedai.atlassian.net/browse/WEB-801) 

[WEB-801]: https://determinedai.atlassian.net/browse/WEB-801?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/hooks/useSettings.ts
Comment (gt2345) [inline webui/react/src/hooks/useSettings.ts:296]: Since we switched to 'subscribe', not sure how should we pass 'shouldPush'

Comment (ClaireNeveu) [reply webui/react/src/hooks/useSettings.ts:296]: So after looking into it, the reason we have `shouldPush` is because we're storing pagination for tables in settings. Specifically we want to pushState instead of replaceState whenever a table's offset changes.

This is really one of those situations where it's obvious useSettings is the wrong abstraction, but changing the table stuff would be quite difficult here. I think the most obvious fix is to add an equivalent of `shouldPush` to the settings config since it's always tied to a specific setting and check if any of the updated values have that.

That said, Thiago did some stuff using state in https://github.com/determined-ai/determined/pull/5775, but I think that's pretty hard to follow.

Comment (gt2345) [reply webui/react/src/hooks/useSettings.ts:296]: I wonder why we used `shouldPush` in the first place, why did we treat `tableOffset` differently?

Comment (ClaireNeveu) [reply webui/react/src/hooks/useSettings.ts:296]: Well, for pagination you _do_ want to push state instead of replacing state. That said I just checked and this feature doesn't actually work properly since we're not listening to the back button and acting appropriately. Let's just remove `shouldPush` for now since the only reason for it to exist isn't working properly. I'll make a ticket to get that working in the future but I don't think that should be part of useSettings when we do it.

File: webui/react/src/hooks/useSettings.ts
Comment (ClaireNeveu) [reply webui/react/src/hooks/useSettings.ts:316]: Another option here: since we're already setting up a subscription to `derivedOb` on [line 146](https://github.com/determined-ai/determined/pull/6202/files#diff-0ac40df189229b81aa5948c8d37e1c96bea9e253bdf2b233fef43661829242f3R146). You could use `usePrevious` to get access to the previous version and do this as a normal react `useEffect` but you'll need to be careful to avoid double-effects due to a dependency on both the current and previous versions in the `useEffect`.

**Review by ClaireNeveu — CHANGES_REQUESTED** [review]
I just saw one minor bug. Other than that this looks good to me but please wait for a review from someone else since I paired with you on this.

File: webui/react/src/hooks/useSettings.ts
Comment (keita-determined) [inline webui/react/src/hooks/useSettings.ts:308]: question: does that have to be `useLayoutEffect`?

File: webui/react/src/hooks/useSettings.ts
Comment (keita-determined) [inline webui/react/src/hooks/useSettings.ts:315]: i think `cur === prev` doesn't check equality https://dmitripavlutin.com/how-to-compare-objects-in-javascript

File: webui/react/src/hooks/useSettings.ts
Comment (keita-determined) [inline webui/react/src/hooks/useSettings.ts:317]: is it safe to use `as` here since `prev` could be `undefined`?

File: webui/react/src/hooks/useSettingsProvider.tsx
Comment (keita-determined) [inline webui/react/src/hooks/useSettingsProvider.tsx:47]: question: how come we need to wrap values with `observable`? to match the props type? it looks unnatural to me.
 @ClaireNeveu Is it the right way to use `micro-observable`?

File: webui/react/src/hooks/useSettings.ts
Comment (keita-determined) [inline webui/react/src/hooks/useSettings.ts:34]: i think we want `shouldPush` or something similar to filter out some query params. For example, if user select some rows in experiment table, `row` params are pushed. However, we want to avoid having `row` in params

**Review by keita-determined — COMMENTED** [review]
I just list what i found. 

---

- [ ] `useSettings` test case is failed, also nice to add `await` before `waitFor`  in [useSettings test cases](https://github.com/determined-ai/determined/blob/master/webui/react/src/hooks/useSettings.test.tsx)

---

- [x] i got an error when `user_web_settings_` table is empty and access to `http://localhost:3000/det/projects/1/experiments` (without any query params) 
  - error message: `useSettings.ts:221 Uncaught (in promise) TypeError: Cannot read properties of undefined 
(reading 'tableLimit')`

---

 - [ ] on [Netlify](https://deploy-preview-6202--determined-ui.netlify.app/projects/1/experiments?archived=true&sortDesc=false&sortKey=SORT_BY_FORKED_FROM&tableLimit=100&tableOffset=100), sorting by `fork` in experiment table doesnt seem working (could be netlify bug)

https://user-images.githubusercontent.com/109113616/226704958-0cde3767-fc49-4f4a-b215-8435da374526.mov


---

- [ ] It's been there but, useSettings doesnt respect query params. is it out of scope in this PR?

https://user-images.githubusercontent.com/109113616/226700686-815f5c29-cf75-4c55-a84d-666b82e1b32c.mov

---

- [x] Spinner doesn't go away after login.

https://user-images.githubusercontent.com/109113616/226701665-598434a8-3843-47e9-b60e-1a7f2b557e6d.mov

---

- [ ] param query isnt updated after switching to different page for the first few times

https://user-images.githubusercontent.com/109113616/226703372-2819a7dc-affd-4958-9417-47093a4b3fcc.mov

---

- [ ] It's a bit slow to update project filtered by users

https://user-images.githubusercontent.com/109113616/226702416-ca852669-4052-45ba-ab8e-88f938a46606.mov

File: webui/react/src/hooks/useSettings.ts
Comment (ClaireNeveu) [reply webui/react/src/hooks/useSettings.ts:315]: Yeah, I suppose we should use `isEqual` for this.

File: webui/react/src/hooks/useSettingsProvider.tsx
Comment (ClaireNeveu) [reply webui/react/src/hooks/useSettingsProvider.tsx:47]: This is correct usage. This value is an observable and we avoid intermediate re-renders that we saw with this being normal react state thereby improving performance.

File: webui/react/src/hooks/useSettings.ts
Comment (ClaireNeveu) [reply webui/react/src/hooks/useSettings.ts:34]: `shouldPush` just determined whether to use replaceState or pushState.

Comment (ClaireNeveu) [reply webui/react/src/hooks/useSettings.ts:34]: https://github.com/determined-ai/determined/blob/master/webui/react/src/hooks/useSettings.ts#L308

Comment (keita-determined) [reply webui/react/src/hooks/useSettings.ts:34]: thanks. so the regression of `row` values in query params comes from other place. i think we talked about this before, and decide not to show `row` and some values in query params.

Comment (ClaireNeveu) [reply webui/react/src/hooks/useSettings.ts:34]: That's part of the config.

File: webui/react/src/hooks/useSettings.ts
Comment (gt2345) [reply webui/react/src/hooks/useSettings.ts:306]: I believe this is to catch any changes in settings before page rendering, think of it as we can fire settings API call and render page at the same time.

File: webui/react/src/hooks/useSettings.ts
Comment (keita-determined) [reply webui/react/src/hooks/useSettings.ts]: the current `isEqual` has a bug too. idk if we need to fix here or different PR 
[typescript playground](https://www.typescriptlang.org/play?removeComments=true&target=99&ts=4.9.4#code/MYewdgzgLgBAlhAogRwK4EMA2MC8MAU6AXDKmANZggDuYANDAEYlmU1gCUJjIImApujC4AfDADeAWABQMOTABO-KKgXCAUgGUA8gDkAdNAVwwAczgAzAJ6EOuHHi17DUY2cs3GHGQF8A3DIyoJCw6LgSMvIwUPzQJADkMdDxdJHyxDDx6PG+gdLB0EzhUrLpCdmppXJJUAk1KbnSQeAQfPz6mCCm+AgoGJiEDF7e0kA)

File: webui/react/src/hooks/useSettings.ts
Comment (keita-determined) [reply webui/react/src/hooks/useSettings.ts:34]: @gt2345 could you add `skipUrlEncoding: true,` for [row](https://github.com/determined-ai/determined/blob/master/webui/react/src/pages/ExperimentList.settings.ts#L135)

File: webui/react/src/hooks/useSettings.ts
Comment (keita-determined) [inline webui/react/src/hooks/useSettings.ts:300]: `s.get(config.storagePath)` and `updates` are both `Settings` object type, so we cant use  `===` here.
also, the code isnt readable due to one liner. i think we should make it readable

File: webui/react/src/hooks/useSettings.ts
Comment (keita-determined) [reply webui/react/src/hooks/useSettings.ts:315]: i think adding `as unknown` is better. idk why ts doesnt show error here. am i missing something?
[ts playground](https://www.typescriptlang.org/play?removeComments=true&target=99&ts=4.9.4#code/C4TwDgpgBAglC8UBKEDGB7ATgEwDwGdhMBLAOwHMAaKQkigPgG4BYAKA1MKgEMAuWKAB8oAV1LYIAMzIRsCUeKkzsbDvnQAbCADoN6cgApuPfLACUq9J0069h491NiA1qXQB3UifNA)

File: webui/react/src/hooks/useSettingsProvider.tsx
Comment (keita-determined) [inline webui/react/src/hooks/useSettingsProvider.tsx:47]: i think we can get rid of `() =>`

**Review by keita-determined — COMMENTED** [review]
there are some comments but overall its working good

note: there are some bugs around useSettings but they have been there before this PR, so i think we can ignore them for now

File: webui/react/src/hooks/useSettings.ts
Comment (gt2345) [reply webui/react/src/hooks/useSettings.ts:34]: What is the context behind changing the behavior here?

Comment (keita-determined) [reply webui/react/src/hooks/useSettings.ts:34]: image

to avoid having this. we used to show it, but decided not to show it. but somehow its visible now

File: webui/react/src/hooks/useSettings.ts
Comment (keita-determined) [reply webui/react/src/hooks/useSettings.ts:308]: i dont think you call the unsubscribe function. the code just returning function, but not calling
https://github.com/BeTomorrow/micro-observables#observablesubscribelistener-value-prevvalue--void

File: webui/react/src/hooks/useSettings.ts
Comment (gt2345) [reply webui/react/src/hooks/useSettings.ts:34]: Just to confirm, we want to hide row for all our tables or just the experiment list table?

Comment (keita-determined) [reply webui/react/src/hooks/useSettings.ts:34]: i think all of them, but let me clarify with Claire. i personally think we dont even need to use useSettings for row tho. 
@ClaireNeveu do we wanna hide all `row` param in URL?

File: webui/react/src/hooks/useSettings.ts
Comment (gt2345) [reply webui/react/src/hooks/useSettings.ts:308]: Not sure if we are on the same page, but if we return a cleanup function,`useEffect` or `useLayoutEffect` should call it for us.
https://react.dev/reference/react/useEffect

Comment (keita-determined) [reply webui/react/src/hooks/useSettings.ts:308]: you are right. i was thinking about different things. nvm

File: webui/react/src/hooks/useSettingsProvider.tsx
Comment (gt2345) [reply webui/react/src/hooks/useSettingsProvider.tsx:47]: I believe we'd want to avoid calling calling functions during initialization
https://react.dev/reference/react/useState#avoiding-recreating-the-initial-state

Comment (keita-determined) [reply webui/react/src/hooks/useSettingsProvider.tsx:47]: i got what you meant. that makes sense. thanks for clarifying

File: webui/react/src/shared/utils/data.ts
Comment (keita-determined) [inline webui/react/src/shared/utils/data.ts:56]: if there's nested object, it wouldn't work but, its fine for now.

**Review by keita-determined — APPROVED** [review]
besides the `row` thing, LGTM

**Pull request merged**
